// expo-cli/build/commands/utils/extractTemplateAppAsync.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createEntryResolver = createEntryResolver;
exports.createFileTransform = createFileTransform;
exports.extractAndPrepareTemplateAppAsync = extractAndPrepareTemplateAppAsync;
exports.extractTemplateAppAsync = extractTemplateAppAsync;
exports.extractTemplateAppFolderAsync = extractTemplateAppFolderAsync;
exports.sanitizeNpmPackageName = sanitizeNpmPackageName;

function _configPlugins() {
  const data = require("@expo/config-plugins");

  _configPlugins = function () {
    return data;
  };

  return data;
}

function _jsonFile() {
  const data = _interopRequireDefault(require("@expo/json-file"));

  _jsonFile = function () {
    return data;
  };

  return data;
}

function _fsExtra() {
  const data = _interopRequireDefault(require("fs-extra"));

  _fsExtra = function () {
    return data;
  };

  return data;
}

function _merge() {
  const data = _interopRequireDefault(require("lodash/merge"));

  _merge = function () {
    return data;
  };

  return data;
}

function _minipass() {
  const data = _interopRequireDefault(require("minipass"));

  _minipass = function () {
    return data;
  };

  return data;
}

function _pacote() {
  const data = _interopRequireDefault(require("pacote"));

  _pacote = function () {
    return data;
  };

  return data;
}

function _path() {
  const data = _interopRequireDefault(require("path"));

  _path = function () {
    return data;
  };

  return data;
}

function _slugify() {
  const data = _interopRequireDefault(require("slugify"));

  _slugify = function () {
    return data;
  };

  return data;
}

function _tar() {
  const data = _interopRequireDefault(require("tar"));

  _tar = function () {
    return data;
  };

  return data;
}

function _xdl() {
  const data = require("xdl");

  _xdl = function () {
    return data;
  };

  return data;
}

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function escapeXMLCharacters(original) {
  const noAmps = original.replace('&', '&amp;');
  const noLt = noAmps.replace('<', '&lt;');
  const noGt = noLt.replace('>', '&gt;');
  const noApos = noGt.replace('"', '\\"');
  return noApos.replace("'", "\\'");
}

class Transformer extends _minipass().default {
  constructor(config, settings) {
    super();
    this.config = config;
    this.settings = settings;

    _defineProperty(this, "data", '');
  }

  write(data) {
    this.data += data;
    return true;
  }

  getNormalizedName() {
    if (['.xml', '.plist'].includes(this.settings.extension)) {
      return escapeXMLCharacters(this.config.name);
    }

    return this.config.name;
  }

  end() {
    const name = this.getNormalizedName();
    const replaced = this.data.replace(/Hello App Display Name/g, name).replace(/HelloWorld/g, _configPlugins().IOSConfig.XcodeUtils.sanitizedName(name)).replace(/helloworld/g, _configPlugins().IOSConfig.XcodeUtils.sanitizedName(name.toLowerCase()));
    super.write(replaced);
    return super.end();
  }

} // Binary files, don't process these (avoid decoding as utf8)


const binaryExtensions = ['.png', '.jar', '.keystore', '.otf', '.ttf'];

function createFileTransform(config) {
  return function transformFile(entry) {
    const extension = _path().default.extname(entry.path);

    if (!binaryExtensions.includes(extension) && config.name) {
      return new Transformer(config, {
        extension
      });
    }

    return undefined;
  };
}
/**
 * Extract a template app to a given file path and clean up any properties left over from npm to
 * prepare it for usage.
 */


async function extractAndPrepareTemplateAppAsync(templateSpec, projectRoot, config) {
  await extractTemplateAppAsync(templateSpec, projectRoot, {
    name: 'name' in config ? config.name : config.expo.name
  });
  const appFile = new (_jsonFile().default)(_path().default.join(projectRoot, 'app.json'));
  const appJson = (0, _merge().default)(await appFile.readAsync(), config);
  await appFile.writeAsync(appJson);
  const packageFile = new (_jsonFile().default)(_path().default.join(projectRoot, 'package.json'));
  const packageJson = await packageFile.readAsync(); // name and version are required for yarn workspaces (monorepos)

  const inputName = 'name' in config ? config.name : config.expo.name;
  packageJson.name = sanitizeNpmPackageName(inputName); // These are metadata fields related to the template package, let's remove them from the package.json.
  // A good place to start

  packageJson.version = '1.0.0';
  packageJson.private = true;
  delete packageJson.description;
  delete packageJson.tags;
  delete packageJson.repository; // pacote adds these, but we don't want them in the package.json of the project.

  delete packageJson._resolved;
  delete packageJson._integrity;
  delete packageJson._from;
  await packageFile.writeAsync(packageJson);
  return projectRoot;
}

function sanitizeNpmPackageName(name) {
  // https://github.com/npm/validate-npm-package-name/#naming-rules
  return applyKnownNpmPackageNameRules(name) || applyKnownNpmPackageNameRules((0, _slugify().default)(name)) || // If nothing is left use 'app' like we do in Xcode projects.
  'app';
}

function applyKnownNpmPackageNameRules(name) {
  // https://github.com/npm/validate-npm-package-name/#naming-rules
  // package name cannot start with '.' or '_'.
  while (/^(\.|_)/.test(name)) {
    name = name.substring(1);
  }

  name = name.toLowerCase().replace(/[^a-zA-Z._\-/@]/g, '');
  return name // .replace(/![a-z0-9-._~]+/g, '')
  // Remove special characters
  .normalize('NFD').replace(/[\u0300-\u036f]/g, '') || null;
}
/**
 * Extract a template app to a given file path.
 */


async function extractTemplateAppAsync(templateSpec, targetPath, config) {
  await _pacote().default.tarball.stream(templateSpec, tarStream => {
    return extractTemplateAppAsyncImpl(targetPath, config, tarStream);
  }, {
    cache: _path().default.join(_xdl().UserSettings.dotExpoHomeDirectory(), 'template-cache')
  });
  return targetPath;
}

async function extractTemplateAppFolderAsync(tarFilePath, targetPath, config) {
  const readStream = _fsExtra().default.createReadStream(tarFilePath);

  await extractTemplateAppAsyncImpl(targetPath, config, readStream);
  return targetPath;
}

function createEntryResolver(name) {
  return entry => {
    if (name) {
      // Rewrite paths for bare workflow
      entry.path = entry.path.replace(/HelloWorld/g, entry.path.includes('android') ? _configPlugins().IOSConfig.XcodeUtils.sanitizedName(name.toLowerCase()) : _configPlugins().IOSConfig.XcodeUtils.sanitizedName(name)).replace(/helloworld/g, _configPlugins().IOSConfig.XcodeUtils.sanitizedName(name).toLowerCase());
    }

    if (entry.type && /^file$/i.test(entry.type) && _path().default.basename(entry.path) === 'gitignore') {
      // Rename `gitignore` because npm ignores files named `.gitignore` when publishing.
      // See: https://github.com/npm/npm/issues/1862
      entry.path = entry.path.replace(/gitignore$/, '.gitignore');
    }
  };
}

async function extractTemplateAppAsyncImpl(targetPath, config, tarStream) {
  await _fsExtra().default.mkdirp(targetPath);
  await new Promise((resolve, reject) => {
    const extractStream = _tar().default.x({
      cwd: targetPath,
      strip: 1,
      // TODO(ville): pending https://github.com/DefinitelyTyped/DefinitelyTyped/pull/36598
      transform: createFileTransform(config),
      onentry: createEntryResolver(config.name)
    });

    tarStream.on('error', reject);
    extractStream.on('error', reject);
    extractStream.on('close', resolve);
    tarStream.pipe(extractStream);
  });
}
